dok_matrix updates

akstrfn opened the pull request:
Hi,

I had problem recently when using pickle on dok_matrix and I saw it was unfixed for quite some time (issue #1188)  hence I decided to fix it. 

But I also added one additional test (not sure if necessary) and made some changes to ``dok_matrix``'s ``__add__`` and ``__radd__`` methods. Namely there are some sort of bound checking that should probably be avoided in these methods. I've also wrote small [script](https://gist.github.com/akstrfn/c3c62e5c798c2fa3211934bec3decc5a) to benchmark some of these things and it looks like that using ``iterkeys`` or ``keys`` is comparably slower. Note I tested only py3 so difference on py2 should be even bigger when using ``keys``.

Please let me know what you think of it and if this looks good I can try to adapt other methods to directly access the data i.e. without bounds checks when useful.

Summary of changes in this PL:
- fix pickle for ``dok_matrix`` in python3 (related issue #1188)
- double for loops changed to ``itertools.product``
- changed some calls to directly call ``dict`` to avoid additional checks
- prevent unintentional usage of ``dok_matrix.update`` since it directly sets the data for ``dok_matrix`` with no checks if the input is valid.
- using generator in combination with ``dict.update`` whenever possible


Cheers,
Aleks

perimosocordiae commented on scipy/sparse/dok.py:
```diff
@@ -427,8 +427,8 @@ def transpose(self, axes=None, copy=False):
-        for key, value in iteritems(self):
-            new[key[1], key[0]] = value
+        for (lhs, rhs), value in iteritems(self):
+            new[rhs, lhs] = value
```
I'd prefer `i` and `j` over `lhs` and `rhs`.

akstrfn replied:
I wasn't sure about this either but since there is a swap between the members I figured better not to use ``i, j`` since they look similar but now looking this ``rhs, lhs`` does not look much better either.

perimosocordiae commented on scipy/sparse/tests/test_base.py:
```diff
@@ -3942,6 +3942,25 @@ def test_ticket1160(self):
         b[:,0] = 0
         assert_(len(b.keys()) == 0, "Unexpected entries in keys")
 
+    def test_copy(self):
```
We already have a test for this: https://github.com/scipy/scipy/blob/master/scipy/sparse/tests/test_base.py#L1784

akstrfn replied:
sorry, totally missed this when I was searching it :smile:

akstrfn commented on scipy/sparse/dok.py:
```diff
@@ -290,21 +291,19 @@ def __setitem__(self, index, x):
-                    if aij != 0:
-                        new[i, j] = aij
+            for i, j in itertools.product(xrange(M), xrange(N)):
+                aij = dict.get(self, (i, j), 0) + other
```
you are right, I totally missed this one

akstrfn commented on scipy/sparse/dok.py:
```diff
@@ -290,21 +291,19 @@ def __setitem__(self, index, x):
+                    new[i, j] = aij
             # new.dtype.char = self.dtype.char
-        elif isinstance(other, dok_matrix):
+        elif isspmatrix_dok(other):
```
To me its also not any clearer but there is a function isspmatrix_dok that, I guess, was supposed to be used for this otherwise it has no purpose?

akstrfn commented on scipy/sparse/dok.py:
```diff
@@ -414,6 +411,9 @@ def __itruediv__(self, other):
             return NotImplemented
 
+    def __reduce__(self):
+        return dict.__reduce__(self)
```
``test_pickle`` that I added fails on python3 for me on 2 ubuntu machines. I just tested and it works in python 2 thought. From debugging I found that it goes to ``spmatrix.__reduce__`` for whatever reason, and that returns an empty dict. I am still not completely sure that this is the right fix though because of all the nuances of subclassing a dict.

perimosocordiae replied:
Ok, I can reproduce the failure on python 3. I would prefer implementing `__getstate__` and `__setstate__` (https://docs.python.org/3.1/library/pickle.html#pickle-state), though.

akstrfn replied:
I tried that but I didn't manage to do it, it was never entering ``__setstate__``. I think because ``dict`` is subclassed and that forces the use of ``__reduce__``? The current solution is just calling ``dict.__reduce__`` reduce instead of ``spmatrix.__reduce__`` in py3 and I assume that's what was already happening in py2. If you have an idea about other implementation let me know so I can give it a shot.

perimosocordiae commented on scipy/sparse/dok.py:
```diff
@@ -93,28 +95,41 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
+
+    def _update(self, data):
+        """ An update method for dict data defined for direct access to
+        `dok_matrix` data. Main purpose is to be used for effient conversion
```
typo: effient -> efficient

perimosocordiae commented on scipy/sparse/dok.py:
```diff
@@ -93,28 +95,41 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
         if axis is not None:
-            raise NotImplementedError("getnnz over an axis is not implemented "
-                                      "for DOK format")
+            raise NotImplementedError("Getnnz over an axis is not implemented "
```
Don't capitalize the method name. Could reorder like: "DOK format doesn't support getnnz() over an axis"

perimosocordiae commented on scipy/sparse/dok.py:
```diff
@@ -140,7 +155,7 @@ def get(self, key, default=0.):
 
     def __getitem__(self, index):
-        """If key=(i,j) is a pair of integers, return the corresponding
+        """ If key=(i,j) is a pair of integers, return the corresponding
```
Don't add space at the start of docstrings (here and elsewhere).

perimosocordiae commented on scipy/sparse/dok.py:
```diff
@@ -351,122 +359,104 @@ def __radd__(self, other):
-        for (i,j),v in iteritems(self):
+        M, N = self.shape
+        n_vecs = other.shape[1]
+        result = np.zeros((M, n_vecs), dtype=upcast(self.dtype, other.dtype))
```
Can clean this a little further:
```python
result_shape = (self.shape[0], other.shape[1])
result_dtype = upcast(self.dtype, other.dtype)
result = np.zeros(result_shape, dtype=result_dtype)
```

perimosocordiae commented on scipy/sparse/dok.py:
```diff
@@ -351,122 +359,104 @@ def __radd__(self, other):
-        for j in range(self.shape[1]):
-            out[0, j] = self[i, j]
-        return out
+        """ Returns a copy of row with index `i` of the matrix as a (1 x n) DOK
```
Returns the i-th row as a (1 x n) DOK matrix.

perimosocordiae commented on scipy/sparse/dok.py:
```diff
@@ -351,122 +359,104 @@ def __radd__(self, other):
-        for i in range(self.shape[0]):
-            out[i, 0] = self[i, j]
-        return out
+        """ Returns a copy of column with index j of the matrix as a (m x 1)
```
Returns the j-th column as a (m x 1) DOK matrix.

perimosocordiae commented on scipy/sparse/tests/test_base.py:
```diff
@@ -1898,6 +1898,26 @@ def test_size_zero_conversions(self):
+            for protocol in range(pickle.HIGHEST_PROTOCOL):
+                sploaded = pickle.loads(pickle.dumps(datsp, protocol=protocol))
+                assert_equal(datsp.shape, sploaded.shape)
+                assert_array_equal(datsp.todense(), sploaded.todense())
```
Use `.toarray()` instead.

perimosocordiae commented on scipy/sparse/tests/test_base.py:
```diff
@@ -1898,6 +1898,26 @@ def test_size_zero_conversions(self):
+                for key, val in datsp.__dict__.items():
+                    if isinstance(val, np.ndarray):
+                        assert_array_equal(val, sploaded.__dict__[key])
+                        continue
```
I'd prefer to drop the `continue` and just use an else-block for the non-ndarray check.

perimosocordiae commented on scipy/sparse/dok.py:
```diff
@@ -351,122 +359,104 @@ def __radd__(self, other):
-            return NotImplemented
+        return NotImplemented
+
+    def __reduce__(self):
```
I'm still not entirely convinced that this is the correct approach, but we can clean it up later if needed. For now, there should be a comment explaining what this is and why's it's needed.

akstrfn replied:
I think one can get it to work with ``__get(set)state__`` but also ``__new__`` and ``__getnewargs__`` would need to be implemented. Reason is that ``__setstate__`` is not called because ``__setitem__`` is called before it and it does not have ``shape`` variable hence everything breaks. If you know some other way to achieve this let me know, but I think this one is reasonably robust since it just redirects work to ``dict``.

akstrfn commented:
@perimosocordiae I've updated the PR to include all your last comments. I've also added one commit where there is a hacky constructor that constructs ``dok_matrix`` from iterator but that can probably be done better. I've ran the benchmarks and some are good and some not, here is the output: [bench.txt](https://github.com/scipy/scipy/files/1195677/bench.txt).

perimosocordiae commented on scipy/sparse/dok.py:
```diff
@@ -93,28 +93,56 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
+        from other spmatrix classes. Has no checking if `data` is valid."""
+        return dict.update(self, data)
+
+
```
The PEP8 checker is complaining about the extra newline here.

perimosocordiae commented:
Your benchmark results look off: too many unrelated timings changed pretty drastically. Are you sure you're doing a fair comparison?

akstrfn commented:
Yes they look very weird, but it was late last night for me to figure out why. I ran them twice and both times got completely weird results. I'll remove iterator constructor and fix that pep8 issue.

pv commented on scipy/sparse/dok.py:
```diff
@@ -497,30 +483,22 @@ def tocsc(self, copy=False):
         if newM < M or newN < N:
             # Remove all elements outside new dimensions
-            for (i, j) in _list(self.keys()):
+            for (i, j) in list(self.keys()):
```
Makes an unnecessary copy on Python 2.

akstrfn replied:
Should be fixed, I also found another place which used `dict.keys()`. How would you go about measuring if a copy is made? I used `%memit` in ipython but did not find any memory difference for dictionary with 10^7 elements. But there is a measurable time difference, `dict.keys()` takes twice as much time compared to `dict.iterkeys()`.

perimosocordiae commented:
I ran the benchmarks locally (using `python runtests.py --bench-compare HEAD~1 sparse\.`). The `before` column is current master (87385d674364fc85709ae4042378eb0d9e1c7b27) and the `after` column is the current state of this PR:
```
    before     after       ratio
+  195.06μs   279.28μs      1.43  sparse.Conversion.time_conversion('coo', 'csc')
+  407.28μs   494.69μs      1.21  sparse.Sum.time_sum_axis1(0.5, 'coo')
+  192.39μs   231.07μs      1.20  sparse.Conversion.time_conversion('coo', 'csr')
+  145.90μs   167.86μs      1.15  sparse.Conversion.time_conversion('csc', 'csr')
+    8.36ms     9.21ms      1.10  sparse.Sum.time_sum_axis0(0.1, 'dia')
+    4.58μs     4.89μs      1.07  sparse.Getset.track_fancy_setitem(1, 'different', 'dok')
+    7.95μs     8.49μs      1.07  sparse.Diagonal.time_diagonal(0.5, 'dia')
+    9.88ms    10.51ms      1.06  sparse.Getset.track_fancy_setitem(10000, 'different', 'csc')
+  241.44μs   255.33μs      1.06  sparse.Densify.time_toarray('coo', 'C')
+   30.39μs    32.00μs      1.05  sparse.NullSlice.time_getrow(0.01, 'csr')
-   47.66μs    45.22μs      0.95  sparse.Getset.time_fancy_getitem(10, 'same', 'dok')
-  718.53μs   676.00μs      0.94  sparse.Conversion.time_conversion('dia', 'coo')
-   16.28ms    15.23ms      0.94  sparse.Conversion.time_conversion('lil', 'dia')
-   48.20μs    45.04μs      0.93  sparse.Getset.time_fancy_getitem(10, 'different', 'dok')
-  713.79μs   666.32μs      0.93  sparse_linalg_onenormest.BenchmarkOneNormEst.time_onenormest(5, 'exact')
-    5.68ms     5.30ms      0.93  sparse.Getset.track_fancy_setitem(1000, 'different', 'csr')
-   10.47ms     9.71ms      0.93  sparse.Matvecs.time_matvecs('dia')
-   12.80ms    11.77ms      0.92  sparse.NullSlice.time_3_rows(0.05, 'csc')
-   30.09ms    25.06ms      0.83  sparse_linalg_lobpcg.Bench.time_sakurai(50, 'lobpcg')
-  103.40μs    76.41μs      0.74  sparse.Getset.time_fancy_getitem(100, 'different', 'dok')
-  106.44μs    75.97μs      0.71  sparse.Getset.time_fancy_getitem(100, 'same', 'dok')
-     1.43s   101.08ms      0.07  sparse.Sum.time_sum_axis0(0.1, 'dok')
-  135.88ms     7.16ms      0.05  sparse.Sum.time_sum_axis0(0.01, 'dok')
```

No significant regressions (COO -> CSC conversions aren't affected by this PR, so that's probably noise) and some very nice speedups on `time_sum_axis0`, which somewhat unintuitively exercises the `_mul_vector` method.

perimosocordiae commented:
Merged. Thanks again, @akstrfn!

akstrfn commented:
@perimosocordiae I ran it in a debugger and ``__rmul__`` is called, hence its the transpose method from ``dok_matrix`` that makes this difference which makes sense since transpose should be much faster now by using ``dict`` builtins and operates directly on data. 
I had to check this because these benchmarks would not make sense otherwise since not much is changed in ``_mul_vector`` except say faster slicing but the difference was too large be explained only by better slicing. Should I submit the  PR with the constructor for ``dok_matrix`` that would take iterators?
Thanks for the feedback and benchmarks!

perimosocordiae commented:
Sure, send the next PR when you're ready. I'd recommend working from a new branch in your fork (that is, not `master`), to avoid git headaches down the road.
